devito/nodes.py

```diff
@@ -484,31 +496,14 @@ def __repr__(self):
     @property
     def _cparameters(self):
         """Generate arguments signature."""
-        cparameters = []
-        for v in self.parameters:
-            if isinstance(v, Dimension):
-                cparameters.append(v.decl)
-            elif v.is_ScalarFunction:
-                cparameters.append(c.Value('const int', v.name))
-            else:
-                cparameters.append(c.Value(c.dtype_to_ctype(v.dtype),
-                                           '*restrict %s_vec' % v.name))
-        return cparameters
+        return [v.decl for v in self.parameters]
```

Comment:
It is related to this and ccasts below. Each argument should be responsible for providing its own ccode, decl and (if applicable) ccast according to the changed approach (which is arguably cleaner than this if-else hierarchy). 

When the DLE creates elemental functions, their arguments are `ScalarFunction` and `TensorFunction` objects that don't support this. Hence they were modified to also provide `Argument` objects so the behaviour can be consistent.